### PR TITLE
set trainer to eval/train mode; wrap `validate` in `torch.inference_mode`

### PR DESCRIPTION
**Pull request by FilippoAiraldi**
In `Trainer.run`, PyTorch modules are never set to train or evaluation mode during training or validation, respectively. This can be an issue if layers such as DropOut are included in any submodules (see [this](https://docs.pytorch.org/docs/stable/notes/autograd.html#evaluation-mode-nn-module-eval)). This commit activates train mode (`self.train()`) or eval mode (`self.eval()`) as needed in `Trainer.run`.

On top of that, calls to `Trainer.validate` in `Trainer.run` have been wrapped in a `torch.inference_mode` context manager to disable gradient computations, hopefully speeding up validation runtime (see [this](https://docs.pytorch.org/docs/stable/notes/autograd.html#inference-mode)).

### CODE REVIEW

**Review by JasperHoffmann — APPROVED** [review]
Looks good!
